cmd/kubeadm/app/cmd/phases/reset/cleanupnode.go

```diff
@@ -188,28 +191,51 @@ func resetConfigDir(configPathDir string, dirsToClean []string, isDryRun bool) {
 }
 
 // CleanDir removes everything in a directory, but not the directory itself
-func CleanDir(filePath string) error {
+func CleanDir(filePath string, dirsToBeRetain []string) error {
 	// If the directory doesn't even exist there's nothing to do, and we do
 	// not consider this an error
 	if _, err := os.Stat(filePath); os.IsNotExist(err) {
 		return nil
 	}
 
-	d, err := os.Open(filePath)
-	if err != nil {
-		return err
-	}
-	defer d.Close()
-	names, err := d.Readdirnames(-1)
-	if err != nil {
-		return err
-	}
-	for _, name := range names {
-		if err = os.RemoveAll(filepath.Join(filePath, name)); err != nil {
+	err := filepath.WalkDir(filePath, func(path string, d os.DirEntry, err error) error {
+		if err != nil {
+			if os.IsNotExist(err) {
+				err = nil // parent was deleted
+			}
+			return err
+		}
+		if IsRetainPath(path, dirsToBeRetain) {
+			return nil
+		}
+		if IsSkipPath(path, dirsToBeRetain) {
+			return filepath.SkipDir
+		}
+		if err = os.RemoveAll(filePath); err != nil {
 			return err
 		}
+		return nil
+	})
+
+	return err
+}
+
+func IsRetainPath(path string, dirsToBeRetain []string) bool {
+	for _, retainPath := range dirsToBeRetain {
+		if strings.Contains(retainPath, path) {
+			return true
+		}
 	}
-	return nil
+	return false
+}
+
+func IsSkipPath(path string, dirsToBeRetain []string) bool {
```

Comment:
IsRetainPath and IsSkipPath seem redundant if we do what my comment above suggests.